src/commands.ts

```diff
@@ -148,11 +148,79 @@ export class Commands
         this._context.outputChannel.show();
     }
 
+    packagePickerChoices(): Promise<{ label, description, pkg }[]>
+    {
+        let choices: { label, description, pkg }[] = [];
+        function addChoice(aPkg: autoproj.Package, ws: autoproj.WorkspaceInfo)
+        {
+            const choice = {
+                label: aPkg.name,
+                description: basename(ws.path),
+                pkg: aPkg
+            }
+            choices.push(choice);
+        }
+
+        return new Promise<{ label, description, pkg }[]>((resolve, reject) => {
+            let pending: any[] = [];
+            this._context.workspaces.forEachWorkspace((ws) => {
+                pending.push(ws.info().then((wsInfo) => {
+                        for (const aPkg of wsInfo.packages) {
+                            if (!this._vscode.workspaceFolders
```

Comment:
Intuitively, I would have created a Map of the available fsPath once outside the promise, and just checked inside whether the folder exists within the promise. The N*M test here feels ... wrong.

Reply:
Do you mean replacing `Array.find()` with `Map.has()` here?

Reply:
The real point would be to calculate the set of available fsPath once and for all outside the promises, and using this inside the promises. Map was more of a suggestion to replace O(N) with O(1)

Reply:
Makes sense.. FYI:

https://jsperf.com/array-iteration-vs-map-keys
https://jsperf.com/array-iteration-vs-object-indexes/4